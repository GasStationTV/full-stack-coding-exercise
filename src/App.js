import React from 'react';
import PropTypes from 'prop-types';

export default class App extends React.Component {
  static propTypes = {
    children: PropTypes.object
  };
  render() {
    return (
      <div>
        <div>This is my app!</div>
        {this.props.children}
      </div>
    );
  }
}
