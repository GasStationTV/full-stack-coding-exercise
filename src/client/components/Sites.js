import React from 'react';
import { Row,Col,ListGroup,ListGroupItem} from 'react-bootstrap';
import FlagsGrid from './FlagsGrid';
import PropTypes from 'prop-types';

const Sites = ({sitesList,toggleFunc,modalCancelFunc,modalSaveFunc,openModalFunc,
  handleChange,handleDateChange,removeAlertFunc,handleAlertDismiss, removeConfirm
  ,flagsList}) => (
  <ListGroup>
    {Object.keys(sitesList).map((key, index)=>{
      return <ListGroupItem key={index}>
        <Row>
          <FlagsGrid site={sitesList[key]} toggleFunc={toggleFunc} siteKey={key}
          modalCancelFunc={modalCancelFunc} modalSaveFunc={modalSaveFunc}
          openModalFunc={openModalFunc} handleChange={handleChange}
          handleDateChange={handleDateChange} removeAlertFunc={removeAlertFunc}
          handleAlertDismiss={handleAlertDismiss} removeConfirm={removeConfirm}
          flagsList={flagsList}/>
        </Row>
      </ListGroupItem>
    })}
  </ListGroup>
)

Sites.propTypes = {
  sitesList: PropTypes.object,
  toggleFunc:PropTypes.func,
  modalCancelFunc:PropTypes.func,
  modalSaveFunc:PropTypes.func,
  openModalFunc:PropTypes.func,
  handleChange:PropTypes.func,
  handleDateChange:PropTypes.func,
  removeAlertFunc:PropTypes.func,
  handleAlertDismiss:PropTypes.func,
  removeConfirm:PropTypes.func,
  flagsList:PropTypes.array
};
export default Sites;
