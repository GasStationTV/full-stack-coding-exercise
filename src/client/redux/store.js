import {createStore,applyMiddleware,combineReducers} from 'redux'
import thunk from 'redux-thunk'
import {routerMiddleware,routerReducer} from 'react-router-redux'
import {hashHistory} from 'react-router'
import {sitesReducer} from './sitesReducer'
import {flagsListReducer} from './flagsListReducer'
const rootReducer = combineReducers({
  sitesReducer,
  flagsListReducer,
  routing: routerReducer
});
const middleware = applyMiddleware(
  routerMiddleware(hashHistory),
  thunk
);
const store = createStore(rootReducer,middleware);
export default store;
