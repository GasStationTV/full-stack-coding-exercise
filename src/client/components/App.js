import React, { Component } from 'react';
import {connect} from 'react-redux';
import Sites from './Sites'
import Navigation from './Navigation'
import {toggleOpenFlag,openModal,cancelModal,handleChange,saveModal,
  removeAlertFunc,handleAlertDismiss,removeConfirm,loadSitesFromAPI,addFlagAPI
  ,editFlagAPI,removeFlagAPI,flagSaveError} from '../redux/siteActions'
import { Row,Alert} from 'react-bootstrap';

class App extends Component{
  constructor(props) {
    super(props);
    this.toggleOpenFlag = this.toggleOpenFlag.bind(this);
    this.openModal=this.openModal.bind(this);
    this.cancelModal=this.cancelModal.bind(this);
    this.handleChange=this.handleChange.bind(this);
    this.saveModal=this.saveModal.bind(this);
    this.handleDateChange=this.handleDateChange.bind(this);
    this.removeAlertFunc=this.removeAlertFunc.bind(this);
    this.handleAlertDismiss=this.handleAlertDismiss.bind(this);
    this.removeConfirm=this.removeConfirm.bind(this);
  }
  componentWillMount(){
    const promise=this.props.dispatch(loadSitesFromAPI());
    promise.then(()=>{},(error)=>{
      console.log("Application failed to load intial data")
    })
  }
  toggleOpenFlag(siteKey){
    this.props.dispatch(toggleOpenFlag(siteKey));
  }
  openModal(siteKey,mode,flagIndex){
    this.props.dispatch(openModal(siteKey,mode,flagIndex));
  }
  cancelModal(siteKey,mode){
    this.props.dispatch(cancelModal(siteKey,mode));
  }
  saveModal(siteKey,mode){
    const candidateFlag=this.props.sitesList[siteKey].flags[this.props.sitesList[siteKey].editFlagIndex]
    let validationError=''
    if(candidateFlag.startDate.length!=0 && candidateFlag.endDate.length!=0){
      if(new Date(candidateFlag.startDate)>=new Date(candidateFlag.endDate)){
        validationError=validationError+' End Date must be after Start Date.'
      }
    }
    if(new Date()>=new Date(candidateFlag.endDate)){
      validationError=validationError+' End Date must be today or in future'
    }
    if(candidateFlag.flagType.length==0){
      validationError=validationError+' Flag Type is required'
    }
    if (validationError!=''){
      this.props.dispatch(flagSaveError(siteKey,validationError));
    }else{
      if(mode=='ADD'){
        this.props.dispatch(addFlagAPI(siteKey,this.props.sitesList));
      }else{
        this.props.dispatch(editFlagAPI(siteKey,this.props.sitesList));
      }
    }
  }
  handleChange(event,siteKey){
    const target = event.target;
    const value = target.type === 'checkbox' ? target.checked : target.value;
    const name = target.name;
    this.props.dispatch(handleChange(siteKey,name,value));
  }
  handleDateChange(date,evt,siteKey,fieldName){
    evt.preventDefault()
    const value=date.toISOString().slice(0, 10);
    this.props.dispatch(handleChange(siteKey,fieldName,value));
  }
  removeAlertFunc(siteKey,flagIndex){
    this.props.dispatch(removeAlertFunc(siteKey,flagIndex));
  }
  handleAlertDismiss(siteKey){
    this.props.dispatch(handleAlertDismiss(siteKey));
  }
  removeConfirm(siteKey){
    this.props.dispatch(removeFlagAPI(siteKey,this.props.sitesList));
  }
  render() {
    return (
    <Row>
      <Row>
        <Navigation/>
      </Row>
      <Row>
        <Sites sitesList={this.props.sitesList} toggleFunc={this.toggleOpenFlag}
        modalCancelFunc={this.cancelModal} modalSaveFunc={this.saveModal} openModalFunc={this.openModal}
         handleChange={this.handleChange} handleDateChange={this.handleDateChange}
         removeAlertFunc={this.removeAlertFunc} handleAlertDismiss={this.handleAlertDismiss}
         removeConfirm={this.removeConfirm} flagsList={this.props.flagsList}/>
      </Row>
    </Row>
    );
  }
}
const mapStateToProps = (state, ownProps) => {
  return {
		sitesList:state.sitesReducer,
    flagsList:state.flagsListReducer
  }
}
const Container = connect(mapStateToProps)(App);
export default Container;
