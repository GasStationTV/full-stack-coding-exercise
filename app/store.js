import { applyMiddleware, createStore } from 'redux';
import sagaMiddleware from 'redux-saga';
import RootReducer from './rootReducer';
import RootSaga from './rootSaga';

export default function Store(initialState) {
  const store = createStore(
    RootReducer,
    initialState,
    applyMiddleware(sagaMiddleware(RootSaga))
  );
  return store;
}
