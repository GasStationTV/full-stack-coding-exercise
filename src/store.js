import { applyMiddleware, combineReducers, compose, createStore } from 'redux';
import thunk from 'redux-thunk';
import { reducer as formReducer } from 'redux-form';
import sitesReducer from './pages/sites/SitesReducer';
import siteReducer from './pages/site/SiteReducer';

const middleware = [thunk];
const enhancers = [];
const reducers = combineReducers({
  form: formReducer,
  sites: sitesReducer,
  site: siteReducer
});

const initialState = {};

const store = createStore(
  reducers,
  initialState,
  compose(applyMiddleware(...middleware), ...enhancers)
);

export default store;
