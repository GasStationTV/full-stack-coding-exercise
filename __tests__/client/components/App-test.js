import React from 'react';
import App from '../../../src/client/components/App'
import Sites from '../../../src/client/components/Sites'
import renderer from 'react-test-renderer';
import store from '../../../src/client/redux/store';
import {Provider} from 'react-redux';
import {shallow} from 'enzyme';
test('App is renderable', () => {
  const component = renderer.create(
    <Provider store={store}>
  		<App/>
  	</Provider>
  );
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test('App DOM validation', () => {
  const appComponent = shallow(
    <Provider store={store}>
  		<App/>
  	</Provider>
  );
  console.log("DEBUG SITES: "+appComponent.find(Sites).length)
  expect(appComponent.find(Sites).length).toBeDefined();
});
