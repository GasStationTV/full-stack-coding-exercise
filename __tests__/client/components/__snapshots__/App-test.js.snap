// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App is renderable 1`] = `
<div
  className="row"
>
  <div
    className="row"
  >
    <nav
      className="navbar navbar-default"
    >
      <div
        className="container"
      >
        <div
          className="navbar-header"
        >
          <a
            className="navbar-brand"
            href="#"
          >
            GSTV - Full Stack Coding Excercise
          </a>
          <button
            className="navbar-toggle collapsed"
            onClick={[Function]}
            type="button"
          >
            <span
              className="sr-only"
            >
              Toggle navigation
            </span>
            <span
              className="icon-bar"
            />
            <span
              className="icon-bar"
            />
            <span
              className="icon-bar"
            />
          </button>
        </div>
      </div>
    </nav>
  </div>
  <div
    className="row"
  >
    <ul
      className="list-group"
    />
  </div>
</div>
`;
